Theme: Generate default ramps at compile time

aduth opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Updates theme package build to generate default ramps at compile time.

It also includes a few other related framework adjustments:

- Adds type-checking to `packages/theme/bin`
   - To support this, revisions were made to the tsconfig validator to consider references in any file matching `tsconfig*.json` in a package
- Allow direct execution of `.ts` files in Node.js using [`esbuild-esm-loader`](https://www.npmjs.com/package/esbuild-esm-loader). This should be a temporary solution until we migrate to a version of Node.js supporting [native type-stripping](https://nodejs.org/en/learn/typescript/run-natively) ([see related discussion](https://github.com/WordPress/gutenberg/pull/71091#issuecomment-3462773907)), and is intended to integrate with our existing ESBuild toolchain.

## Why?

The default experience will use these ramps. For default values, we can know ahead of time the results of the built ramps, so we can precompute these and avoid the added runtime cost.

## How?

Build script now generates default ramps.

## Testing Instructions

Validate no regressions in behavior of behavior of theme provider:

1. Run `npm run storybook:start`
2. Go to http://localhost:50240/?path=/docs/design-system-theme-provider--docs
3. Observe that colors generate as expected

Verify that building theme package succeeds and produces no local changes:

1. Run `npm run --workspace @wordpress/theme build`
2. Observe successful run
3. Run `git status`
4. Observe clean working branch

aduth committed:
Update tsconfig validator to allow multi-config packages

aduth committed:
Theme: Generate default ramps at compile time

aduth commented on bin/validate-tsconfig.mjs:
```diff
@@ -3,10 +3,9 @@
 import glob from 'glob';
 import { dirname, basename } from 'path';
-import stripJsonComments from 'strip-json-comments';
+import JSONC from 'jsonc-parser';
```
tsconfig files are JSONC files, so we may as well use a JSONC parser than trying to transform the string into something we can shove into `JSON.parse`. This is the library that TypeScript itself maintains and [uses internally](https://github.com/microsoft/TypeScript/blob/6fd2874d0dc24aeca92887bdadf6a63429ce1f1b/scripts/build/utils.mjs#L5).

aduth commented on package.json:
```diff
@@ -109,6 +109,7 @@
 		"deep-freeze": "0.0.1",
 		"equivalent-key-map": "0.2.2",
 		"esbuild": "0.25.10",
+		"esbuild-esm-loader": "0.3.3",
```
This is a [package](https://www.npmjs.com/package/esbuild-esm-loader) that I maintain and have used in a number of other projects. It itself is pretty light-weight and delegates most of the underlying work to [Node's module customization hooks](https://nodejs.org/docs/latest/api/module.html#customization-hooks) and ESBuild.

youknowriad replied:
Why is it added to the top level package instead of the theme package where it's used?

aduth replied:
I'm considering it a `devDependencies` and as far as I recall we define all of those in the root `package.json`?

youknowriad replied:
It is possible that how things started, but I'm thinking that ideally going forward, each package should define its own dependencies both prod and dev. Unless the dependency is used outside packages. (bin for instance).

Not a blocker for this PR, something we can iterate on later.

aduth replied:
I think that makes sense, especially here where it is a devDependency (relevant for compile-time), but only for this package, and not repo-wide. Happy to make the change.

aduth replied:
The tooling is set up to disallow it, so I'll have to make some revisions there.

```
✖ prefer-no-devDependencies - node: devDependencies - devDependencies should not be defined
```

aduth replied:
Updated in d02d9e072e (allow devDependencies) and 72ad4d18e6 (shift to package devDependency).

aduth committed:
Allow package-level devDependencies

Makes it clearer which dependency is relevant for which packages. Dev dependencies which are "repo-wide" (i.e. used in root-level scripts) should still be defined in the root package.json

aduth committed:
Move esbuild-esm-loader as devDependency of theme package

See: https://github.com/WordPress/gutenberg/pull/72847#discussion_r2479580862

jsnajdr commented on packages/theme/src/color-ramps/lib/default-ramps.ts:
```diff
@@ -0,0 +1,173 @@
+> = {
+	bg: {
+		ramp: {
+			surface2: { color: '#f8f8f8', warning: false },
```
In near future, we could eliminate the `warning` field. It's ever used only by Storybook to mark colors that fail to meet the contrast requirements.

In practice, we always adjust the seed color so that the entire ramp fits within the 0.0-1.0 lightness range, with all contrast requirements met. The `warning` flag is always `false` unless you're doing some debugging.

aduth replied:
> In near future, we could eliminate the `warning` field. It's ever used only by Storybook to mark colors that fail to meet the contrast requirements.
> 
> In practice, we always adjust the seed color so that the entire ramp fits within the 0.0-1.0 lightness range, with all contrast requirements met. The `warning` flag is always `false` unless you're doing some debugging.

Yeah I think we could do a few things to condense the shape of this object so it doesn't take up so much space.

Even if we want `warning` as is, we could maybe have it as an (optional) adjacent key of the ramp so that each of these ramp values is just the string hex value rather than an object.

e.g.

```js
bg: {
  ramp: {
    surface2: '#f8f8f8',
    /* ... */
  },
  direction: 'darker',
  /* warnings: [ 'surface2' ] */
}
```

aduth replied:
I created a follow-up issue at #72931
